aqt/archives.py

```diff
@@ -38,6 +49,358 @@ def __init__(self, version, target, arch, os_name):
         self.os_name = os_name
 
 
+class ListCommand:
+    """Encapsulate all parts of the `aqt list` command"""
+
+    # Inner helper classes
+    class Versions:
+        def __init__(self, it_of_it: Iterable[Tuple[int, Iterable[Version]]]):
+            self.versions: List[List[Version]] = [
+                list(versions_iterator) for _, versions_iterator in it_of_it
+            ]
+
+        def __str__(self):
+            return "\n".join(
```

Comment:
IMHO, it is not good habit to include "\n" in __str__

Reply:
Ok, I can change this.

While we're on this subject, maybe we should discuss exactly how the output for this feature should be formatted.

This feature prints different kinds of data in different ways:

* Versions of qt: Currently, we print all qt versions in ascending order, separated by spaces, with each patch number on a different line. For example:

      6.0.0 6.0.1 6.0.2 6.0.3 6.0.4
      6.1.0 6.1.1 6.1.2
      6.2.0

* Architectures, modules, extensions, and targets: These are all printed on a single line, with each item separated by a space character.
* Tools: These are all printed on separate lines. At first, I did this because many of the tool names are quite long, and it was easier for me to read this way.  In retrospect, I think that it's very strange that the tools aren't all printed on the same line like everything else. Also, it's strange that all of the tools begin with the prefix `tools_`. As a user, I would expect `aqt list tools` to print `ifw cmake qtcreator` instead of `tools_ifw tools_cmake tools_qtcreator`. It's a waste of horizontal space.

Is there anything on this list that you would like changed?

Reply:
I think `__str__` for human-readability, and pretty printing for CLI are different.
When logging object state to a log file in future, we will use `str()` in style `logger.log(obj)` that produce a line feed on single entry in a log file. It is not good.

Is it difficult define a named function for purpose of pretty print, and use it from CLI?

Reply:
Sorry I did not make it more clear in my last message, but I intend to add a ‘pretty_print’ -type function and use it instead of ‘str’. That is easy to do. I just wanted to give you an opportunity to change what that ‘pretty_print’ function prints.

Reply:
It is ok as same as current `__str__()` doing.

> Tools: These are all printed on separate lines. At first, I did this because many of the tool names are quite long, and it was easier for me to read this way. In retrospect, I think that it's very strange that the tools aren't all printed on the same line like everything else. Also, it's strange that all of the tools begin with the prefix tools_. As a user, I would expect aqt list tools to print ifw cmake qtcreator instead of tools_ifw tools_cmake tools_qtcreator. It's a waste of horizontal space.

I agree with it. It is ok for me about current output.

Reply:
Ok, it should be fixed now.